File: solutions/__116_max_product.go
package main

// 又是看题解的一天
func maxProduct(nums []int) int {
	max := func(a, b int) int {
		if a > b {
			return a
		}
		return b
	}
	min := func(a, b int) int {
		if a < b {
			return a
		}
		return b
	}

	n := len(nums)
	rtv := nums[0]
	maxF, minF := rtv, rtv
	for i := 1; i < n; i++ {
		maxF, minF = max(maxF*nums[i], max(minF*nums[i], nums[i])), min(maxF*nums[i], min(minF*nums[i], nums[i]))
		rtv = max(rtv, maxF)
	}
	return rtv
}

File: solutions/__116_max_product.cpp
//
// Created by peanut on 2021/8/16.
//

#include <bits/stdc++.h>

using namespace std;

int maxProduct(vector<int> &nums) {
    int maxF = nums[0], minF = nums[0], ans = nums[0];
    for (int i = 1; i < nums.size(); ++i) {
        int mx = maxF, mn = minF;
        maxF = max(mx * nums[i], max(nums[i], mn * nums[i]));
        minF = min(mn * nums[i], min(nums[i], mx * nums[i]));
        ans = max(maxF, ans);
    }
    return ans;
}
